### PR TITLE
Allow cudf.concat to process a dictionary of frames #15115

### LINKED ISSUE #15115: [FEA] Allow `cudf.concat` to process a dictionary of frames / add `keys=`
In pandas, you can pass a dictionary of dataframes to construct a MultiIndex dataframe. This is especially useful if the values contain frames that have the same columns, in which case the keys can help differentiate them.

Right now, the following code:
```python
d = {
    'first': cudf.DataFrame({'A': [1, 2], 'B': [3, 4]}),
    'second': cudf.DataFrame({'A': [5, 6], 'B': [7, 8]}),
}

cudf.concat(d, axis=1)
```
raises the error
```none
TypeError: cannot concatenate object of type <class 'str'>
```
If we try to construct a dataframe from the values of `d`:
```python
cudf.concat(d.values(), axis=1)
```
we get:
```none
NotImplementedError: A Column with duplicate name found: A, cuDF doesn't support having multiple columns with same names yet.
```

If we try to pass keys:
```none
cudf.concat(d.values(), keys=d.keys(), axis=1)
```
we get
```none
TypeError: concat() got an unexpected keyword argument 'keys'
```

Right now, I can do one of two things to get around this issue:
- Mangle the column names of each dataframe beforehand (in a loop) to make them unique and concatenate them. Afterwards, work the columns to make it MultiIndex. 
- Concatenate at the cupy level, create MultiIndex from scratch and create a new dataframe object from these.

In pandas, we could do `pd.concat(d, axis=1)` and call it a day. I wish the same was possible in cudf.

### PR DESCRIPTION
**Pull request by amanlai**
Closes #15115.

Unlike `pandas.concat`, `cudf.concat` doesn't work with a dictionary of objects. The following code raises an error.

```python
d = {
    'first': cudf.DataFrame({'A': [1, 2], 'B': [3, 4]}),
    'second': cudf.DataFrame({'A': [5, 6], 'B': [7, 8]}),
}

cudf.concat(d, axis=1)
cudf.concat(d)
```

This commit resolves this issue.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/tests/test_concat.py [hunk 1]
```diff
@@ -1921,3 +1921,21 @@ def test_concat_mixed_list_types_error(s1, s2):
...
     with pytest.raises(NotImplementedError):
         cudf.concat([s1, s2], ignore_index=True)
+
+
+def test_horizontal_concat_dictionary():
+
+    d = {
+        'first': cudf.DataFrame({'A': [1, 2], 'B': [3, 4]}),
+        'second': cudf.DataFrame({'A': [5, 6], 'B': [7, 8]}),
+    }
```
Comment (shwina) [inline python/cudf/cudf/tests/test_concat.py:1931]: Can we parametrize to include a few additional test cases?

- A single entry in `d`
- A third entry in `d` with just 1 key-value pair

**Comment by shwina** [comment]
Looks like the PR is failing style checks. Can you run locally:

```
pre-commit run --all
```

and fix up any style issues reported?

File: python/cudf/cudf/tests/test_concat.py (hunk 1 above)
Comment (amanlai) [reply python/cudf/cudf/tests/test_concat.py:1931]: Aren't single entry in `d` and a third entry in `d` with just 1 key-value pair the same thing? Are you saying there should be a test for `d = {'first': cudf.DataFrame({'A': [1, 2], 'B': [3, 4]})}`?

File: python/cudf/cudf/tests/test_concat.py [hunk 2]
```diff
@@ -1921,3 +1921,21 @@ def test_concat_mixed_list_types_error(s1, s2):
 
     with pytest.raises(NotImplementedError):
         cudf.concat([s1, s2], ignore_index=True)
+
+
+def test_horizontal_concat_dictionary():
```
[the remark this answers is not in the document]
Comment (amanlai) [reply python/cudf/cudf/tests/test_concat.py:1926]: I can add tests for these.

**Comment by amanlai** [comment]
@shwina I screwed up and deleted the branch by mistake which I think closed the PR. How do I start this up again?

**Comment by amanlai** [comment]
@shwina just to clarify, do I create a new pull request?

**Comment by shwina** [comment]
Yes I think that's best

**Comment by amanlai** [comment]
@shwina I ended up making a new pull request at #15160.

**Pull request closed without merging**